docs/middleware.md

```diff
@@ -0,0 +1,218 @@
+
+- It takes each `CR` that arrives to a topic.
+- It can then do something to the `CR` or run any needed code.
+- Then it passes the `CR` to be processed by the specific stream.
+- Once the `CR` is processed by the stream, the chain is "completed".
+- If there is code after the `self.next_call(cr)` then it will be executed.
+
+Kstreams `Middleware` have the following protocol:
+
+::: kstreams.middleware.middleware.Middleware
+
+!!! note
+    The `__call__` method can return anything so previous calls can use the returned value. Make sure that the line `return await self.next_call(cr)` is in your method
+
+!!! warning
+    Middlewares only work with the new [Dependency Injection approach](https://kpn.github.io/kstreams/stream/#dependency-injection-and-typing)
+
+## Creating a middleware
+
+To create a middleware you have to create a class that inherits from `BaseMiddleware`. Then, the method `async def __call__` must be defined. Let's consider that we want to save the CR to `elastic` before it is processed:
+
+```python
+import typing
+
+from kstreams import ConsumerRecord, middleware
+
+async def save_to_elastic(cr: ConsumerRecord) -> None:
+    ...
+
+
+class ElasticMiddleware(middleware.BaseMiddleware):
+    async def __call__(self, cr: ConsumerRecord) -> typing.Any:
+        # save to elastic before calling the next
+        await save_to_elastic(cr)
+
+        # the next call could be another middleware
+        return await self.next_call(cr)
+```
+
+Then, we have to include the middleware. It can be per `stream` or per application level (`stream_engine`):
+
+=== "Include middleware in stream"
+    ```python
+    from kstreams import ConsumerRecord, middleware
+
+    from .engine import stream_engine
+
+
+    middlewares = [middleware.MiddlewareFactory(ElasticMiddleware)]
+
+    @stream_engine.stream("kstreams-topic", middlewares=middlewares)
+        async def processor(cr: ConsumerRecord):
+            ...
+    ```
+=== "Include middleware at application level"
+    ```python
+    from kstreams import ConsumerRecord, middleware, create
+
+    middlewares = [middleware.MiddlewareFactory(ElasticMiddleware)]
+    
+    stream_engine = kstreams.create_engine(
+        title="my-stream-engine",
+        middlewares=middlewares
+    )
+    ```
+
+!!! note
+    The example middlewares make sense to include it at the application level, but other ones like Dead Letter Queue (DLQ) make sense to add them at stream level
+
+!!! note
+    The `Middleware` concept also applies for `async generators` (yield from a stream)
+
+## Adding extra configuration to middlewares
+
+If you want to provide extra configuration to middleware you should override the __init__ method, ensuring that is contains the keywargs `next_call`, `send` and `stream`, then any remaining are optional `keywargs`.
```

Comment:
This line also mentions "override", but the example is implementing the protocol. Should the example inherit from `BaseMiddleware`?

Reply:
Yes, it is possible and then we can use `super().__init__(next_call=next_call, send=send, stream=stream)`